test/e2e/node_feature_discovery.go

```diff
@@ -506,6 +627,51 @@ func waitForNfdNodeLabels(cli clientset.Interface, expected map[string]string) e
 	return err
 }
 
+// waitForNfdNodeTaints waits for node to be tainted as expected.
+func waitForNfdNodeTaints(cli clientset.Interface, expected []corev1.Taint) error {
+	poll := func() error {
+		nodes, err := getNonControlPlaneNodes(cli)
+		if err != nil {
+			return err
+		}
+		for _, node := range nodes {
+			taints, err := nfdTaints(node.Spec.Taints, node)
+			if err != nil {
+				return fmt.Errorf("failed to fetch nfd owned taints for node: %s", node.Name)
+			}
+			if !cmp.Equal(expected, taints) {
+				return fmt.Errorf("node %q taints do not match expected, diff (expected vs. received): %s", node.Name, cmp.Diff(expected, taints))
+			}
+		}
+		return nil
+	}
+
+	// Simple and stupid re-try loop
+	var err error
+	for retry := 0; retry < 3; retry++ {
+		if err = poll(); err == nil {
+			return nil
+		}
+		time.Sleep(10 * time.Second)
+	}
+	return err
+}
+
+// nfdTaints returns taints that are owned by the nfd.
+func nfdTaints(taints []corev1.Taint, node corev1.Node) ([]corev1.Taint, error) {
+	// De-serialize the taints annotation into corev1.Taint type for comparision below.
+	var err error
+	nfdTaints := []corev1.Taint{}
+	if val, ok := node.Annotations[nfdv1alpha1.NodeTaintsAnnotation]; ok {
```

Comment:
You don't actually use the `taints` arg for anything even if that should be the thing you're looking at. I suggest forget about the annotation altogether (as that might be wrong), just filter all taints whose key starts with `nfd.node.kubernetes.io`(we know we use that key for the test taints)

Reply:
Good point. Fixed